File: PromoStudio.Common/Models/TemplateScript.cs
﻿using System;
using System.Collections.Generic;
using PromoStudio.Common.Enumerations;

namespace PromoStudio.Common.Models
{
    public class TemplateScript
    {
        private List<TemplateScriptItem> _scriptItems = new List<TemplateScriptItem>();

        public long pk_TemplateScriptId { get; set; }
        public sbyte fk_TemplateScriptStatusId { get; set; }
        public int? fk_OrganizationId { get; set; }
        public short fk_StoryboardItemType { get; set; }
        public string Name { get; set; }
        public string Description { get; set; }
        public string ProjectFilePath { get; set; }
        public string ThumbnailFilePath { get; set; }
        public string PreviewFilePath { get; set; }
        public string RenderCompName { get; set; }
        public string RenderPreviewCompName { get; set; }
        public double? RenderCompStartTime { get; set; }
        public double? RenderCompEndTime { get; set; }
        public DateTime DateCreated { get; set; }
        public DateTime? DateUpdated { get; set; }

        public List<TemplateScriptItem> Items
        {
            get { return _scriptItems; }
            set { _scriptItems = value; }
        }

        public TemplateScriptStatus Status
        {
            get { return (TemplateScriptStatus) fk_TemplateScriptStatusId; }
            set { fk_TemplateScriptStatusId = (sbyte) value; }
        }

        public StoryboardItemType StoryboardType
        {
            get { return (StoryboardItemType) fk_StoryboardItemType; }
            set { fk_StoryboardItemType = (short) value; }
        }

        public dynamic ToPoco()
        {
            return new
            {
                pk_TemplateScriptId,
                fk_TemplateScriptStatusId,
                fk_OrganizationId,
                fk_StoryboardItemType,
                Name,
                Description,
                ProjectFilePath,
                ThumbnailFilePath,
                PreviewFilePath,
                RenderCompName,
                RenderPreviewCompName,
                RenderCompStartTime,
                RenderCompEndTime,
                DateCreated,
                DateUpdated
            };
        }
    }
}

File: PromoStudio.Web/Scripts/models/templateScript.js
/// <reference path="../vsdoc/require.js" />
/// <reference path="../vsdoc/knockout-3.0.0.debug.js" />

"use strict";

define(["models/templateScriptItem", "knockout"], function (templateScriptItem, ko) {
    function ctor (data) {
        var self = this;
        data = data || {};

        self.pk_TemplateScriptId = ko.observable(data.pk_TemplateScriptId || null);
        self.fk_TemplateScriptStatusId = ko.observable(data.fk_TemplateScriptStatusId || null);
        self.fk_OrganizationId = ko.observable(data.fk_OrganizationId || null);
        self.Name = ko.observable(data.Name || null);
        self.Description = ko.observable(data.Description || null);
        self.ProjectFilePath = ko.observable(data.ProjectFilePath || null);
        self.ThumbnailFilePath = ko.observable(data.ThumbnailFilePath || null);
        self.PreviewFilePath = ko.observable(data.PreviewFilePath || null);
        self.RenderCompName = ko.observable(data.RenderCompName || null);
        self.RenderPreviewCompName = ko.observable(data.RenderPreviewCompName || null);
        self.RenderCompStartTime = ko.observable(data.RenderCompStartTime || null);
        self.RenderCompEndTime = ko.observable(data.RenderCompEndTime || null);
        
        self.Items = ko.observableArray([]);

        self.LoadItems = function (items) {
            var i, item;
            items = items || [];

            for (i = 0; i < items.length; i++) {
                item = items[i];
                items[i] = new templateScriptItem(item);
            }

            self.Items(items);
        };
        self.LoadItems(data.Items);
    }

    return ctor;
});